mdtk/pytorch_datasets.py

```diff
@@ -1 +1,283 @@
 """classes to use in conjunction with pytorch dataloaders"""
+from torch.utils.data import Dataset
+import tqdm
+import torch
+import random
+import pandas as pd
+import numpy as np
+
+#TODO: probably want to move everything but Pytorch dataset objects out of here
+
+# Convenience function...
+def diff_pd(df1, df2):
+    """Identify differences between two pandas DataFrames"""
+    assert (df1.columns == df2.columns).all(), \
+        "DataFrame column names are different"
+    if any(df1.dtypes != df2.dtypes):
+        "Data Types are different, trying to convert"
+        df2 = df2.astype(df1.dtypes)
+    if df1.equals(df2):
+        return None
+    else:
+        # need to account for np.nan != np.nan returning True
+        diff_mask = (df1 != df2) & ~(df1.isnull() & df2.isnull())
+        ne_stacked = diff_mask.stack()
+        changed = ne_stacked[ne_stacked]
+        changed.index.names = ['id', 'col']
+        difference_locations = np.where(diff_mask)
+        changed_from = df1.values[difference_locations]
+        changed_to = df2.values[difference_locations]
+        return pd.DataFrame({'from': changed_from, 'to': changed_to},
+                            index=changed.index)
+
+# TODO: later can auto detect vocab from corpus if necessary
+#       I'm doing things this way just for ability to change things
+#       later with ease
```

Comment:
I'm not sure it would be reasonably possible to detect the vocab later. We can read and find all of the used vocab symbols (as ints), but it will be much more complicated to figure out exactly min/max pitch, and time_increment.

We should write the params for this later.

Reply:
Oh, nevermind! I see the csv contain strings, and the loader converts to ints.

Reply:
Yeah, this is what the initial vocab does - just defines all possible strings expected in commands. However, this is not very efficient - there may be some strings not present in the data, and if we just want to read the data and find all possible tokens (no rounding of ms to nearest 40ms or whatever), then you can do that relatively easily. That's all I mean by this comment.

Reply:
The rounding I implemented actually now pre-quantizes everything before writing to the csv, so converting back out without quantization is impossible, if we keep it. But I made the convert metadata to cmd csv its own function so people can easily create unquantized vocab that way.